### ISSUE TITLE
The CIS pod only matters with one namespace

### ISSUE THREAD

**Issue by kylinsoong** [F5Networks/k8s-bigip-ctlr#1337]
#### Setup Details
 CIS Version : 2.X      
 Build: f5networks/k8s-bigip-ctlr:2.0.0   
 BIGIP Version: Big IP 13.1.3 
 AS3 Version: 3.19.0      
 Agent Mode: AS3/CCCL       
 Orchestration: K8S/OSCP      
 Orchestration Version:       
 Pool Mode: Cluster    
 Additional Setup details: <Platform/CNI Plugins/ cluster nodes/ etc>

#### Description  
  
After upgrade the cis pod version to 2.0.0, the cis only can monitor on one namespace, if more than one namespace exists, there will be a error throw:

~~~
2020/06/04 23:46:55 [ERROR] [AS3] Invalid user-defined cfgMap with name: app-as3-1-partition-2-vs and namespace f5-test002
2020/06/04 23:47:41 [ERROR] [AS3] Invalid user-defined cfgMap with name: app-as3-1-partition-2-vs and namespace f5-test002
2020/06/04 23:47:41 [ERROR] [AS3] Invalid user-defined cfgMap with name: app-as3-1-partition-2-vs and namespace f5-test002
2020/06/04 23:47:41 [ERROR] [AS3] Invalid user-defined cfgMap with name: app-as3-1-partition-2-vs and namespace f5-test002
2020/06/04 23:48:11 [ERROR] [AS3] Invalid user-defined cfgMap with name: app-as3-1-partition-2-vs and namespace f5-test002
2020/06/04 23:48:11 [ERROR] [AS3] Invalid user-defined cfgMap with name: app-as3-1-partition-2-vs and namespace f5-test002
2020/06/04 23:48:41 [ERROR] [AS3] Invalid user-defined cfgMap with name: app-as3-1-partition-2-vs and namespace f5-test002
2020/06/04 23:48:41 [ERROR] [AS3] Invalid user-defined cfgMap with name: app-as3-1-partition-2-vs and namespace f5-test002
2020/06/04 23:49:11 [ERROR] [AS3] Invalid user-defined cfgMap with name: app-as3-1-partition-2-vs and namespace f5-test002
~~~

#### Steps To Reproduce
1) Install the CIS
~~~
apiVersion: apps/v1
kind: Deployment
metadata:
  name: k8s-bigip-ctlr
  namespace: kube-system
  labels:
    app: k8s-bigip-ctlr
spec:
  replicas: 1
  selector:
    matchLabels:
      app: k8s-bigip-ctlr
  template:
    metadata:
      name: k8s-bigip-ctlr
      labels:
        app: k8s-bigip-ctlr
    spec:
      serviceAccountName: bigip-ctlr
      containers:
        - name: k8s-bigip-ctlr
          image: "f5networks/k8s-bigip-ctlr:2.0.0"
          env:
            - name: BIGIP_USERNAME
              valueFrom:
                secretKeyRef:
                  name: bigip-login
                  key: username
            - name: BIGIP_PASSWORD
              valueFrom:
                secretKeyRef:
                  name: bigip-login
                  key: password
          command: ["/app/bin/k8s-bigip-ctlr"]
          args: [
            "--bigip-username=$(BIGIP_USERNAME)",
            "--bigip-password=$(BIGIP_PASSWORD)",
            "--bigip-url=https://10.1.1.245:443",
            "--insecure=true",
            "--bigip-partition=k8s",
            "--pool-member-type=cluster",
            "--vs-snat-pool-name=TEST_SNAT_Pool",
            ]
~~~
2) Deploy VS 1(will success)
~~~
kind: ConfigMap
apiVersion: v1
metadata:
  name: app-as3
  namespace: f5-test001
  labels:
    f5type: virtual-server
    as3: "true"
data:
  template: |
    {
      "class": "AS3",
      "action": "deploy",
      "persist": true,
      "declaration": {
        "class": "ADC",
        "schemaVersion": "3.18.0",
        "id": "123abc",
        "label": "f5_test001",
        "remark": "HTTP application",
        "f5_test001": {
          "class": "Tenant",
          "f5_test001": {
            "class": "Application",
            "template": "generic",
            "app_svc_vs": {
              "class": "Service_HTTP",
              "persistenceMethods": [ "cookie" ],
              "virtualAddresses": [
                "10.1.10.60"
              ],
              "virtualPort": 80,
              "pool": "app_svc_pool"
            },
            "app_svc_pool": {
              "class": "Pool",
              "monitors": [
                "tcp"
              ],
              "members": [
              {
                "servicePort": 80,
                "serverAddresses": []
              }
              ]
            }
          }
        }
      }
    }
~~~
3) Deploy 2nd VS(Will failed with error log)
~~~
kind: ConfigMap
apiVersion: v1
metadata:
  name: app-as3-1-partition-2-vs
  namespace: f5-test002
  labels:
    f5type: virtual-server
    as3: "true"
data:
  template: |
    {
      "class": "AS3",
      "action": "deploy",
      "persist": true,
      "declaration": {
        "class": "ADC",
        "schemaVersion": "3.11.0",
        "id": "one_partition_two_vs",
        "label": "f5_test002",
        "remark": "HTTP application",
        "f5_test002": {
          "class": "Tenant",
          "f5_test002_1": {
            "class": "Application",
            "template": "generic",
            "app_1_svc_vs": {
              "class": "Service_HTTP",
              "virtualAddresses": [
                "10.1.10.61"
              ],
              "virtualPort": 80,
              "pool": "app_1_svc_pool"
            },
            "app_1_svc_pool": {
              "class": "Pool",
              "monitors": [
                "tcp"
              ],
              "members": [
              {
                "servicePort": 80,
                "serverAddresses": []
              }
              ]
            }
          },
           "f5_test002_2": {
            "class": "Application",
            "template": "generic",
            "app_2_svc_vs": {
              "class": "Service_HTTP",
              "virtualAddresses": [
                "10.1.10.62"
              ],
              "virtualPort": 80,
              "pool": "app_2_svc_pool"
            },
            "app_2_svc_pool": {
              "class": "Pool",
              "monitors": [
                "tcp"
              ],
              "members": [
              {
                "servicePort": 80,
                "serverAddresses": []
              }
              ]
            }
          }
        }
      }
    }
~~~

NOTE: due to the CIS can only monitor on one namespace, so the namespace isolation(--namespace, --namespace-label) also not work.

**Comment by cisbotctlr** [comment]
cisbot will assign the issue to one of the devs. 
 @devs, use /jira for internal tracking.

**Comment by kylinsoong** [comment]
@cisbotctlr give me the /jira link please

**Comment by mdditt2000** [comment]
@kylinsoong CIS can monitor multiple namespace.
```
 args: 
            - "--bigip-username=$(BIGIP_USERNAME)"
            - "--bigip-password=$(BIGIP_PASSWORD)"
            - "--bigip-url=192.168.200.92"
            - "--bigip-partition=k8szoneb"
            - "--namespace=test2-icni-1"
            - "--namespace=nat46"
            - "--pool-member-type=cluster"
            - "--flannel-name=fl-vxlan"
            - "--log-level=DEBUG"
            - "--insecure=true"
            - "--manage-ingress=false"
            - "--manage-routes=false"
            - "--manage-configmaps=true"
            - "--custom-resource-mode=false"
            - "--http-listen-address=0.0.0.0:8080"
            - "--as3-validation=true"
            - "--log-as3-response=true"
          command: 
```
Example: https://github.com/mdditt2000/kubernetes-1-18/blob/master/cis%202.0/user-defined-configmap/big-ip-92-cluster/cis-deployment/f5-cluster-deployment.yaml

**Comment by AbhishekPy** [comment]
> @cisbotctlr @agupta49 give me the /jira link please

Jira will be created only once the issue is triaged and accepted.

**Comment by cisbotctlr** [comment]
> @cisbotctlr @agupta49 give me the /jira link please

@kylinsoong Jira will be created only once the issue is triaged and accepted by dev team.

**Comment by kylinsoong** [comment]
@mdditt2000 Have you verified that config?, I just tested in my environment, it's still doesn't work.

Also, the issue come out due to the upgrade the CIS version from 1.14 to 2.0, In customer environment, there are 23 namespaces in K8S, and 24 partitions in BIG-IP, prior to the upgrade one CIS POD can monitor multiple namspaces, after upgrade, the CIS POD only watch on one namespace.

The change also affect the namespace isolation, the --namespace-label are used to isolation, after upgrade, it also not works.

**Comment by trinaths** [comment]
@kylinsoong With agent AS3, only 1 configMap across all watched namespaces will work. That means, though you configure CIS to watch multiple namespaces or use namespace-labels, you can only have ONE single configMap for all your apps. So, combine confgMap `app-as3` and `app-as3-1-partition-2-vs` to a single one and try. For Using Namespace Labels, there is an issue in 2.0 resolved and will be available with 2.1. Refer to upcoming release [Release-notes](https://github.com/F5Networks/k8s-bigip-ctlr/blob/master/docs/RELEASE-NOTES.rst#bug-fixes).

Refer docs for more info on how Userdefined-as3-configMaps works.

[1] https://clouddocs.f5.com/containers/latest/userguide/config-map.html
[2] https://github.com/F5Networks/k8s-bigip-ctlr/tree/master/docs/_static/config_examples/configmap/user-defined-configmap

**Comment by kylinsoong** [comment]
@trinaths Thanks for replying, does this is the designed behavior start from 2.0? in previous version, it's not behave like this, in previous version also use the agent AS3, but it can

1. multiple configmap in different namespace
2. multiple configmap in single namespace

Actually the customer has multiple configmap in different namespace in their current environment.

> So, combine confgMap app-as3 and app-as3-1-partition-2-vs to a single one and try.
Where should put the configmap? note that in my reproduction, `app-as3` in namespace 001, `pp-as3-1-partition-2-vs` in namespace 002.

**Comment by trinaths** [comment]
@kylinsoong - From CIS 1.14 too the same is the behaviour. The behaviour change is, in CIS 1.14, the default agent is CCCL and in 2.0 its AS3. Though CIS is watching `n` number of namespaces, only one User defined AS3 ConfigMap is processed. For more info, please refer -  https://clouddocs.f5.com/containers/latest/userguide/config-map.html

**Comment by myf5** [comment]
Customers usually use NS to mappting to project. If CIS only processes 1 AS3-Confgimap, that means different project owners need control/edit the same configmap, this may cause global issue if someone put wrong configmap. Also this is not match to RBAC mind.

**Comment by kylinsoong** [comment]
@trinaths Thanks for updating.

> The behaviour change is, in CIS 1.14, the default agent is CCCL and in 2.0 its AS3. Though CIS is watching n number of namespaces, only one User defined AS3 ConfigMap is processed. 

The key here is the big inconsistent with 1.14, 

- In 1.14, the customer has lots of configmaps in different namespace, each namespace can have one or more configmap, CIS POD can watch on all these configmaps
- In 2.0, Only one configmap can be use, so after the upgrade, customer has to convert all configmaps in different  namespace to one configmap? also which namespace is the best place to hold the configmap, can one configmap can config other namespace's service?

**Comment by kylinsoong** [comment]
I have finished some test and get the following founding:

* The global configmap should not put on kube-system namespace
* The cis.f5.com annotation should unique across all service, If put all in one configmap, the cis.f5.com annotation on each service should not be equal, eg, the following 2 services can not discover successfully due to they have same `cis.f5.com/as3-pool` annotation

*Service one:*
~~~
apiVersion: v1
kind: Service
metadata:
  labels:
    app: demo-as3-app
    cis.f5.com/as3-tenant: "f5_test001"
    cis.f5.com/as3-app: "f5_test001"
    cis.f5.com/as3-pool: "app_svc_pool"
  name: app-svc
  namespace: f5-test001
~~~

*Service two:*
~~~
apiVersion: v1
kind: Service
metadata:
  labels:
    app: app
    cis.f5.com/as3-tenant: f5_test019
    cis.f5.com/as3-app: f5_test019
    cis.f5.com/as3-pool: app_svc_pool
  name: app-svc
  namespace: f5-test019
~~~

* I have finished the test that put all configuration on one User defined AS3 ConfigMap, each namespace's services put on one tenant, there are 3 tenants in my test, also can add more tenant if need, now my concern is **does add new tenant is affect the exist tenant?** 

eg, if there already has 3 tenant in configmap reference 3 partition in BIG-IP, and VS under these partitions are online in production, now modify configmap add 4th tenant, does this operation affect exist online VS? **If add 4th tenant cause production VS redeploy, this will cause business inconsistent not acceptable**

@trinaths Could you confirm my above concern? put all in one whether has possibility to cause business inconsistent or not?

---

In CIS 1.14, the default agent is CCCL, I use the following args in bigip-ctrl that can work AS3 template in configmap, so I get a conclusion that the default CCCL agent in CIS 1.14 support AS3 template,
~~~
args: [
            "--bigip-username=$(BIGIP_USERNAME)",
            "--bigip-password=$(BIGIP_PASSWORD)",
            "--bigip-url=https://10.1.1.245:443",
            "--insecure=true",
            "--bigip-partition=k8s",
            "--pool-member-type=cluster",
            "--vs-snat-pool-name=TEST_SNAT_Pool"
    ]
~~~

In CIS 2.0, the default agent is AS3, with the conclusion above, if we force CIS 2.0 to use CCCL, eg add --agent args
~~~
args: [
            "--bigip-username=$(BIGIP_USERNAME)",
            "--bigip-password=$(BIGIP_PASSWORD)",
            "--bigip-url=https://10.1.1.245:443",
            "--insecure=true",
            "--bigip-partition=k8s",
            "--pool-member-type=cluster",
            "--vs-snat-pool-name=TEST_SNAT_Pool",
            "--agent=cccl"
    ]
~~~
this should be work as CIS 1.14 default cccl agent, but **the test results is it doesn't work**, My question here is **why CIS 1.14 with default cccl agent can work with AS3 template, and force CIS 2.0 use cccl agent not work?**

**Comment by mdditt2000** [comment]
@myf5 @kylinsoong you might want to switch to CRDs https://github.com/mdditt2000/kubernetes-1-18/blob/master/cis%202.0/crd/CustomResource.md 

CRDs introduced in aplha CIS 2.0. Enhancements coming in CIS 2.1 which include all TLS options.

**Comment by mdditt2000** [comment]
@kylinsoong this is regular AS3 behavior 

Question -  add new tenant is affect the exist tenant?

eg, if there already has 3 tenant in configmap reference 3 partition in BIG-IP, and VS under these partitions are online in production, now modify configmap add 4th tenant, does this operation affect exist online VS? If add 4th tenant cause production VS redeploy, this will cause business inconsistent not acceptable

Please review - What is the difference between AS3 Selective and Complete updates and why is this important?
https://clouddocs.f5.com/products/extensions/f5-appsvcs-extension/latest/userguide/faq.html

Selective is the default behavior in which tenants not explicitly referenced in declaration rePOSTing **are NOT modified.**

**Comment by kylinsoong** [comment]
Thanks @mdditt2000 , I can have a whole picture now, seems CRD works on k8s 1.18+, not sure whether CRD is compatible with K8s prior to version 1.18. Thanks for answer my questions.

Below is a summary of my experience with CC/CIS:

1. At the early start age, we use CC with iApp(current still have run in production), there are very confused log message, then we switch to use the CCCL with CIS in new project, but this mode lack of some key F5 configuration, like cookie based persistence, customize tcp idle time out, enable XXF, etc,
2. Then we switch to AS3 with CIS, it works fine when the number of namespaces less then 15, in our UAT environment , there are 25 namespaces, add a new VS to F5 need more then 10 mins, also the F5 Management RAM are always 100% consumed.
3. So we try to use `--namespace-label` to isolate namespace, to use multiple CIS Pod to work parallel, this can increase the performance, VS deployment become faster, but the service auto discover features missed, #1311 is addressing this missing.
4. Then 2.0 come out, the big change are only allow one user defined configmap(1.14 allow multiple ), also 2.0 not complete solve service auto discover issue, refer to #1311

**Comment by mdditt2000** [comment]
Going to be resolved in CIS 2.1 - Internal PM Jira task CONTCNTR-1945 - Remove "--userdefined-as3-declaration" from the deployment arguments. And cleanup all the dependent code.

Duplicate of https://github.com/F5Networks/k8s-bigip-ctlr/issues/1311

**Comment by mdditt2000** [comment]
Resolved in CIS 2.1
